File: 707_software_patterns/assn/03/prog/src/ValidChildren.cpp
#include "ValidChildren.hpp"

ValidChildren::ValidChildren(std::string _thisElement)
    : thisElement(_thisElement), _canHaveText(false) {}

ValidChildren::~ValidChildren() {}

std::string ValidChildren::getThisElement() {
    return thisElement;
}

bool ValidChildren::canHaveText() {
    return _canHaveText;
}

void ValidChildren::setCanHaveText(bool __canHaveText) {
    _canHaveText = __canHaveText;
}

void ValidChildren::addValidChild(const std::string& child, bool isAttribute) {

    if (childIsValid(child, isAttribute)) {
        return;
    }

    validChildren.push_back(child);
    childIsAttribute.push_back(isAttribute);
}

bool ValidChildren::childIsValid(const std::string& child, bool isAttribute) {

    for (unsigned int i = 0; i < validChildren.size(); i++) {

        if (childIsAttribute[i] == isAttribute &&
            validChildren[i].compare(child) == 0) {
            return true;
        }
    }

    return false;
}

File: 707_software_patterns/assn/03/prog/sol/edu/jhu/apl/patterns_class/ValidChildren.java
package edu.jhu.apl.patterns_class;

public class ValidChildren
{
	private String				thisElement		= null;	// A value of null represents Document.
	private java.util.Vector<String>	validChildren		= new java.util.Vector<String>();
	private java.util.Vector<Boolean>	childIsAttribute	= new java.util.Vector<Boolean>();
	private boolean				_canHaveText		= false;

	public ValidChildren(String thisElement)		{ this.thisElement = thisElement; }

	public String	getThisElement()			{ return thisElement; }
	public boolean	canHaveText()				{ return _canHaveText; }
	public void	setCanHaveText(boolean _canHaveText)	{ this._canHaveText = _canHaveText; }

	public void	addValidChild(String child, boolean isAttribute)
	{
		if (childIsValid(child, isAttribute))
			return;

		validChildren.add(child);
		childIsAttribute.add(new Boolean(isAttribute));
	}

	public boolean	childIsValid(String child, boolean isAttribute)
	{
		for (int i = 0; i < validChildren.size(); i++)
			if (childIsAttribute.elementAt(i).booleanValue() == isAttribute &&
			  validChildren.elementAt(i).compareTo(child) == 0)
				return true;

		return false;
	}

	//
	// For XML validation completeness, ValidChildren should also have methods for verifying the validity of attribute values
	// as commented in original version of XMLValidator.
	//
}
